api/pkg/transformer/pipeline/environment.go

```diff
@@ -44,6 +44,10 @@ func (e *Environment) Preprocess(ctx context.Context, rawRequest types.Payload,
 	e.symbolRegistry.SetRawRequestHeaders(rawRequestHeaders)
 	e.SetOutput(rawRequest)
 
+	if !e.IsPreprocessOpExist() {
```

Comment:
What's the purpose of this condition here? Won't `e.compiledPipeline.Preprocess(ctx, e)` return immediately if there is no preprocess operation?

Reply:
previously this one is called in `handler.go` to return immediately without doing `Unmarshaling` the request/response if there is no preprocess/postprocess . I move it to `environment.go` to handle case where user doesn't specify preprocess or postprocess ops, so when publish the prediction log the rawRequest can be fetched from `symbolRegistry` but I just realized by moving it to `environment.go` it defeats the purpose of that condition (to reduce the unmarshal when there is no preprocess or postprocess operation)